File: crm_backend/app/model/product.go
package model

import (
	"gorm.io/gorm"
)

// Enum for Size Variant Types
type SizeVariantType string

const (
	Alpha   SizeVariantType = "Alpha"
	Numeric SizeVariantType = "Numeric"
	Free    SizeVariantType = "Free"
)

type Product struct {
	gorm.Model
	SKU       string  `gorm:"size:100;not null;unique" json:"sku"`
	Name      string  `gorm:"size:255;not null" json:"name"`
	Status    bool    `gorm:"default:true" json:"status"` // Active/inactive product status
	Desc      string  `gorm:"size:255;not null;" json:"description"`
	MRP       float64 `gorm:"not null" json:"mrp"`        // Maximum Retail Price
	CostPrice float64 `gorm:"not null" json:"cost_price"` // Custom field name in JSON response
	SellPrice float64 `gorm:"not null" json:"sell_price"`

	// Foreign keys to the new tables
	CategoryID uint            `json:"-"` // Foreign key to the Category table
	Category   ProductCategory `gorm:"foreignKey:CategoryID" json:"category"`

	FitID uint       `json:"-"` // Foreign key to the Fit table
	Fit   ProductFit `gorm:"foreignKey:FitID" json:"fit"`

	VariantID uint           `json:"-"` // Foreign key to the Variant table
	Variant   ProductVariant `gorm:"foreignKey:VariantID" json:"variant"`

	ColorID uint         `json:"-"` // Foreign key to the Color table
	Color   ProductColor `gorm:"foreignKey:ColorID" json:"color"`

	FabricID uint          `json:"-"` // Foreign key to the Fabric table
	Fabric   ProductFabric `gorm:"foreignKey:FabricID" json:"fabric"`

	SleeveID uint          `json:"-"` // Foreign key to the Sleeve table
	Sleeve   ProductSleeve `gorm:"foreignKey:SleeveID" json:"sleeve"`

	GenderID uint          `json:"-"` // Foreign key to the Gender table
	Gender   ProductGender `gorm:"foreignKey:GenderID" json:"gender"`

	SizeVariant SizeVariantType `gorm:"type:enum('Alpha', 'Numeric', 'Free'); not null" json:"size_variant"` // Enum for size variant type

	SourceID uint          `json:"-"` // Foreign key to the Source table
	Source   ProductSource `gorm:"foreignKey:SourceID" json:"source"`

	GRNProducts []GRNProduct   `gorm:"foreignKey:ProductID"` // Explicit relationship to GRNProduct table
	Images      []ProductImage `gorm:"foreignKey:ProductID" json:"images"`
	ImageFileID string         `gorm:"size:100" json:"image_file_id"`

	CreatedByID uint `json:"-"`                               // Foreign key for the user who created the product
	CreatedBy   User `gorm:"foreignKey:CreatedByID" json:"-"` // Reference to User table

	LastModifiedByID uint `json:"-"`                                    // Foreign key for the user who last modified the product
	LastModifiedBy   User `gorm:"foreignKey:LastModifiedByID" json:"-"` // Reference to User table
}

type ProductImage struct {
	gorm.Model
	ProductID uint    `json:"product_id"` // Foreign key to the Product table
	Product   Product `gorm:"foreignKey:ProductID" json:"-"`
	FileID    string  `gorm:"size:255;not null" json:"file_id"` // SeaweedFS file ID
	URL       string  `json:"url"`
	Position  int     `json:"position"` // Ordering of images for the product
}

type ProductProperty struct {
	Name        string `gorm:"size:100;not null;unique" json:"name"`
	CreatedByID uint   `json:"-"` // Foreign key to the user who created this property
	CreatedBy   User   `gorm:"foreignKey:CreatedByID" json:"-"`
}

// ProductCategory model
type ProductCategory struct {
	gorm.Model
	ProductProperty
}

// ProductFit model
type ProductFit struct {
	gorm.Model
	ProductProperty
}

// ProductVariant model
type ProductVariant struct {
	gorm.Model
	ProductProperty
}

// ProductColor model
type ProductColor struct {
	gorm.Model
	ProductProperty
}

// ProductFabric model
type ProductFabric struct {
	gorm.Model
	ProductProperty
}

// ProductSleeve model
type ProductSleeve struct {
	gorm.Model
	ProductProperty
}

// ProductGender model
type ProductGender struct {
	gorm.Model
	ProductProperty
}

// ProductSource model
type ProductSource struct {
	gorm.Model
	ProductProperty
}

File: material-ui-cra-ts/src/model/product.ts
import User from "./user";
// api request
export interface ProductFormDataRequest {
	id: number;
	sku: string;
	name: string;
	status: boolean;
	mrp: number;
	cost_price: number;
	sell_price: number;
	category_id: number;
	fit_id: number;
	variant_id: number;
	color_id: number;
	fabric_id: number;
	sleeve_id: number;
	gender_id: number;
	size_variant_id: number;
	source_id: number;
}

export interface ProductFormDataResponse {
	success: boolean;
	product: ProductFormDataRequest;
}

export interface ProductPropertyAddResponse {
	success: boolean;
	property_name: string;
	property: ProductProperty;
}

// api response
export interface FilteredProductsResponse {
	"page": number,
	"page_size": number,
	"total_items": number,
	"total_pages": number,
	products: Array<Product>,
}

export interface FiltersResponse {
	filters: Array<FilterResponse>
}

export interface FilterResponse {
	name: string;
	type: string;
	metadata: any;
}

export interface ProductPropertiesResponse {
	properties: {
		categories: ProductProperty[];
		colors: ProductProperty[];
		fabrics: ProductProperty[];
		fits: ProductProperty[];
		genders: ProductProperty[];
		size_variants: ProductProperty[];
		sleeves: ProductProperty[];
		sources: ProductProperty[];
		variants: ProductProperty[];
	}
}



// product model
export default class Product {
	ID: number;
	sku: string;
	name: string;
	status: boolean;
	mrp: number;
	cost_price: number;
	sell_price: number;
	category: ProductProperty;
	fit: ProductProperty;
	variant: ProductProperty;
	color: ProductProperty;
	fabric: ProductProperty;
	sleeve: ProductProperty;
	gender: ProductProperty;
	size_variant: ProductProperty;
	source: ProductProperty;
	createdBy: User;
	lastModifiedBy: User;

	constructor(
		ID: number,
		sku: string = "",
		name: string = "",
		status: boolean = false,
		mrp: number = 0,
		cost_price: number = 0,
		sell_price: number = 0,
		category: ProductProperty,
		fit: ProductProperty,
		variant: ProductProperty,
		color: ProductProperty,
		fabric: ProductProperty,
		sleeve: ProductProperty,
		gender: ProductProperty,
		size_variant: ProductProperty,
		source: ProductProperty,
		createdBy: User,
		lastModifiedBy: User
	) {
		this.ID = ID;
		this.sku = sku;
		this.name = name;
		this.status = status;
		this.mrp = mrp;
		this.cost_price = cost_price;
		this.sell_price = sell_price;
		this.category = category;
		this.fit = fit;
		this.variant = variant;
		this.color = color;
		this.fabric = fabric;
		this.sleeve = sleeve;
		this.gender = gender;
		this.size_variant = size_variant;
		this.source = source;
		this.createdBy = createdBy;
		this.lastModifiedBy = lastModifiedBy;
	}
}

export class ProductProperty {
	ID: number;
	name: string;
	// createdBy: User;
	isChecked: boolean;

	constructor(id: number, name: string = "", isChecked: boolean = false) {
		this.ID = id;
		this.name = name;
		// this.createdBy = createdBy;
		this.isChecked = isChecked;
	}
}


// filters
export class ProductFilter {
	name: string;
	filterType: string;
	isSelected: boolean;

	constructor(name: string, filterType: string, isSelected = false) {
		this.name = name;
		this.filterType = filterType;
		this.isSelected = isSelected;
	}
}

export class RangeFilter extends ProductFilter {
	min: number;
	max: number;

	constructor(name: string, filterType = "range", isSelected = false, min = 0, max = 90000) {
		super(name, filterType, isSelected);
		this.min = min;
		this.max = max;
	}
}

export class LikeFilter extends ProductFilter {
	content: string;

	constructor(name: string, filterType = "range", isSelected = false, content: string) {
		super(name, filterType, isSelected);
		this.content = content;
	}
}

export class ListFilter extends ProductFilter {
	values: Array<ProductProperty>;

	constructor(name: string, filterType = "range", isSelected = false, values: Array<ProductProperty>) {
		super(name, filterType, isSelected);
		this.values = values;
	}
}

